Sources/core/objects/Gun.cpp

```diff
@@ -0,0 +1,105 @@
+#include "Gun.h"
+#include "Tile.h"
+
+Gun::Gun(size_t id) : Item(id)
+{
+    SetSprite("icons/guns.dmi");
+    ammunition_ = 1;
+    max_ammunition_ = 1;
+}
+
+bool Gun::UseAmmo()
+{
+    if (ammunition_ <= 0)
+    {
+        ammunition_ = 0;
+        return false;
+    }
+    --ammunition_;
+    return true;
+}
+
+void Gun::ShootImpl(VDir target, const std::string& sound, const std::string& projectile_type, const std::string& casing_type)
+{
+    id_ptr_on<CubeTile> tile = GetOwner()->GetOwner();
+    if (tile.valid())
+    {	
+        if (UseAmmo())
+        { 
+            auto projectile = Create<Projectile>(projectile_type,tile);
+            projectile->MakeMovementPattern(target,GetOwner());
+            PlaySoundIfVisible(sound,tile.ret_id());
+            if(!casing_type.empty())
+            {
+                Dir dir = GetRand() % 4;
+                id_ptr_on<Item> bc =Create<Item>(casing_type,tile.ret_id());
+                bc->Rotate(dir);
+            } 	
+        }
+        else
+        {
+            PlaySoundIfVisible("empty.ogg",tile.ret_id()); 
+            // the *click* text from ss13
+        }
+    }	
+}
+
+void Gun::Shoot(VDir target)
+{
+}
+
+bool Gun::AddAmmo()
+{
+    if (ammunition_ < max_ammunition_)
+    {
+        ++ammunition_;
+        return true;
+    }
+    return false;
+}
+void Gun::AttackBy(id_ptr_on<Item> item)
+{
+    if (id_ptr_on<AmmunitionBox> box = item)
+    {
+        if(box->CheckBullets())
+        {
+            if(AddAmmo())
+            {
+		        box->RemoveBullet();
+                return;
+            }
+        }
+        // message which tells the player that the box is empty 
+    }
```

Comment:
They will be using different ammunition boxes but I didn't added the right boxes yet because that wasn't my main goal so I think the function will change a bit.